### PR TITLE
feat(main): RPC Router and Server implementation

### PR DESCRIPTION
**Pull request by kjdelisle**
connected to https://github.com/strongloop/loopback.io/pull/523

Currently no tests in place for RPC Server (router has most of the actual logic)

### CODE REVIEW

File: test/controllers/greet.controller.test.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import {expect} from '@loopback/testlab';
+
+describe('greet.controller', () => {
+  const controller = new GreetController();
+  describe('basicHello', () => {
+    it('returns greetings for a name', () => {
+      expect(controller.basicHello({})).to.equal('Hello, World!');
+    });
+
+    it('returns greetings for the world without valid input', () => {
```
Comment (shimks) [inline test/controllers/greet.controller.test.ts:12]: `it('returns greetings with valid input, ...`

File: test/servers/rpc-router.test.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  let request: express.Request;
+  let response: express.Response;
+  // tslint:disable-next-line:no-any
+  let controller: any;
+
+  beforeEach(setupFakes);
+  it('routes to existing controller and method', async () => {
+    await router.routeRequest(request, response);
+    const stub = response.send as sinon.SinonStub;
+    expect(stub.called);
```
Comment (bajtos) [inline test/servers/rpc-router.test.ts:18]: Have you seen this assertion to fail? I believe it's a no-op:

```
> expect(false)
Assertion {
  obj: false,
  anyOne: false,
  negate: false,
  params: { actual: false } }
```

I think you wanted to say `expect(stub.called).to.be.true()` but please don't do that - the assertion failure will be rather unhelpful. Use [sinon assertions](http://sinonjs.org/releases/v1.17.6/assertions/) instead:

```ts
sinon.assert.called(stub);
```

We tried to integrate should with sinon via https://github.com/shouldjs/sinon, but it did not work :(

File: test/servers/rpc-router.test.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+  function getController(rtr: RPCRouter) {
+    const stub = sinon.stub(rtr, 'getController');
+    stub.resolves(new FakeController());
+    return stub;
+  }
+
+  function getRequest(req?: Partial<express.Request>) {
+    const reqt = Object.assign(
+      <express.Request>{
```
Comment (bajtos) [inline test/servers/rpc-router.test.ts:69]: Uh oh, this is hacky and may easily break when express starts to require on another request property that's not mocked out here, or calls any of the ReadableStream APIs. Have you considered using [Shot](https://www.npmjs.com/package/shot) instead? We are shipping it as a part of our testlab, see [packages/testlab/src/shot.ts](https://github.com/strongloop/loopback-next/blob/df1c8799f681af30680a390f5d8081290c9047e7/packages/testlab/src/shot.ts)

File: test/servers/rpc-router.test.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+          },
+        },
+      },
+      req,
+    );
+    return reqt;
+  }
+
+  function getResponse(res?: Partial<express.Response>) {
+    const resp = <express.Response>{};
```
Comment (bajtos) [inline test/servers/rpc-router.test.ts:84]: Ditto here, please consider using Shot's Response mock instead.

File: test/servers/rpc-router.test.ts (hunk 2 above)
Comment (bajtos) [reply test/servers/rpc-router.test.ts:18]: Please fix other places using `expect(stub.called);` too

File: test/servers/rpc-router.test.ts (hunk 3 above)
Comment (kjdelisle) [reply test/servers/rpc-router.test.ts:69]: @bajtos Would you be able to provide an example of how to use our testlab exports of Shot? I've noticed that we're not actually using it the way Shot's Github README advertises, and I'm not certain how I would use it to mock the handling of express requests and responses (given that it's for `http.ServerRequest` and `http.ServerResponse`).

I understand that the hard cast here could break if something leverages a function or property that I haven't faked, but I've gone out of my way to make sure that nothing calls functions on the request or response object in the rpc-router code above and beyond what's faked here.

Comment (kjdelisle) [reply test/servers/rpc-router.test.ts:69]: Technically speaking, no instance of an express server ever begins listening on an external port; these tests are essentially integration tests.

Comment (bajtos) [reply test/servers/rpc-router.test.ts:69]: > Would you be able to provide an example of how to use our testlab exports of Shot?

Have you tried to search for the places in loopback-next that are using it? There are quire few test files leveraging `mockResponse`, see https://github.com/strongloop/loopback-next/search?utf8=%E2%9C%93&q=shot&type=

Looks like we need to improve the docs for this part of testlab! I created an issue - see https://github.com/strongloop/loopback-next/issues/760

**Pull request merged**